sdk/pubkey/src/lib.rs

```diff
@@ -14,68 +14,77 @@ use pinocchio::pubkey::{Pubkey, MAX_SEEDS, PDA_MARKER};
-/// the expected resulting address.
+/// This function avoids the cost of the `create_program_address` syscall
+/// (`1500` compute units) by directly computing the derived address
+/// calculating the hash of the seeds, bump and program id using the
+/// `sol_sha256` syscall.
 ///
 /// # Important
 ///
-/// This function differs from [`pinocchio::pubkey::create_program_address`] in that it
-/// does not perform a validation to ensure that the derived address is a valid
+/// This function differs from [`pinocchio::pubkey::create_program_address`] in that
+/// it does not perform a validation to ensure that the derived address is a valid
 /// (off-curve) program derived address. It is intended for use in cases where the
 /// seeds, bump, and program id are known to be valid, and the caller wants to derive
 /// the address without incurring the cost of the `create_program_address` syscall.
-pub fn derive_address<const N: usize>(seeds: &[&[u8]; N], bump: u8, program_id: &Pubkey) -> Pubkey {
+pub fn derive_address<const N: usize>(
+    seeds: &[&[u8]; N],
+    bump: Option<u8>,
+    program_id: &Pubkey,
+) -> Pubkey {
     const {
-        assert!(
-            N <= MAX_SEEDS,
-            "number of seeds must be less than MAX_SEEDS"
-        );
+        assert!(N < MAX_SEEDS, "number of seeds must be less than MAX_SEEDS");
     }
 
     const UNINIT: MaybeUninit<&[u8]> = MaybeUninit::<&[u8]>::uninit();
-    let mut data = [UNINIT; MAX_SEEDS + 3];
+    let mut data = [UNINIT; MAX_SEEDS + 2];
     let mut i = 0;
 
     while i < N {
-        // SAFETY: `data` is guanranteed to have enough space for `N` seeds,
+        // SAFETY: `data` is guaranteed to have enough space for `N` seeds,
         // so `i` will always be within bounds.
         unsafe {
             data.get_unchecked_mut(i).write(seeds.get_unchecked(i));
         }
         i += 1;
     }
 
-    let bump = [bump];
+    // TODO: replace this with `as_slice` when the MSRV is upgraded
+    // to `1.84.0+`.
+    let bump_seed = [bump.unwrap_or_default()];
 
-    // SAFETY: `data` is guaranteed to have enough space for `MAX_SEEDS + 3`
+    // SAFETY: `data` is guaranteed to have enough space for `MAX_SEEDS + 2`
     // elements, and `MAX_SEEDS` is as large as `N`.
     unsafe {
-        data.get_unchecked_mut(i).write(bump.as_ref());
-        data.get_unchecked_mut(i + 1).write(program_id.as_ref());
-        data.get_unchecked_mut(i + 2).write(PDA_MARKER.as_ref());
+        if bump.is_some() {
+            data.get_unchecked_mut(i).write(&bump_seed);
+            i += 1;
+        }
+        data.get_unchecked_mut(i).write(program_id.as_ref());
+        data.get_unchecked_mut(i + 1).write(PDA_MARKER.as_ref());
     }
 
     #[cfg(target_os = "solana")]
     {
         let mut pda = MaybeUninit::<[u8; 32]>::uninit();
 
-        // SAFETY: `data` has `i + 3` elements initialized.
+        // SAFETY: `data` has `N + 2` elements initialized.
```

Comment:
Should this say `i + 2`?

```suggestion
        // SAFETY: `data` has `i + 2` elements initialized.
```